### DISCUSSION TITLE
제페토 멀티플레이(Room State)

### DISCUSSION THREAD

**Discussion by yoonjoohwan** [naverz/zepeto-studio-kor#394] [5. Q&A]
가이드 4-3 에서  State 에 변화가 있으면 OnStateChange 가 자동으로 실행된다고되어있습니다
혹시 State 에 변화를 Client (ZepetoScript 에 Ts 를 이용해서)에서 줄 수 있는 방법이 있을까요?

**Comment by minjujuu** [comment]
https://github.com/naverz/zepeto-multiplay-example
위 멀티플레이 예제 참고하시면 좋을 것 같아요!

클라이언트에서 room.Send()를 통해 서버로 상태 변경에 관한 메시지를 보내고,
서버에서 메시지를 받으면 state를 변경하도록 하는 방식을 사용해보시면 어떨까요?

클라이언트 코드에서 메시지를 보내는 예시>
```
private SendState(state: CharacterState) {
        const data = new RoomData();
        data.Add("state", state);
        this.room.Send("onChangedState", data.GetObject());
    }
```

서버 코드에서 메시지를 받는 예시>
```
        this.onMessage("onChangedState", (client, message) => {
            const player = this.state.players.get(client.sessionId);
            // state가 변경되어 클라이언트의 OnStateChange가 호출됨
            player.state = message.state;
        });
```
그리고 변경된 state에 대한 값은 
OnStateChange(state:State, isFirst:boolean) { } 에서
첫번째 매개변수인 state를 통해 접근하실 수 있습니다

**Reply by yoonjoohwan** [reply]
그렇다면 멀티플레이 환경에서 서로에게 보이고 움직임이 공유되는 객체를 만들고 싶다면 RoomState 에 저장하여 구현하면 되는걸까요?
플레이어 같은 느낌으로 무언가를 만들어보고싶어서요

**Reply by minjujuu** [reply]
넵, 말씀주신 부분도 위 예제에 구현되어 있어서 프로젝트를 참고해보시면 좋을 듯 한데요~
서로의 움직임이 동기화되는 코드에 대해서는 다음 내용 위주로 확인하시면 좋을 것 같습니다.
간단하게만 적은거라 실제 구현하실 때에는 처리해주어야 할 부분이 더 있을 거에요

1. schemas - Schema Types에 Vector3, Transform, Player 존재, Room state에  Player의 MapSchema 존재 
![image](https://user-images.githubusercontent.com/79794414/154385978-6fb3b021-549b-4c65-9343-ae7fb21c17f7.png)
2. 클라이언트 - 일정 시간 간격으로 room.Send()를 통해 내 캐릭터 transform을 server로 전송 
3. 서버 - onMessage를 통해 클라이언트에서 보낸 메시지를 받아 player state에 변경된 transform 반영
4. 클라이언트 - OnStateChange에서 player.OnChange()(* player의 state가 갱신될 때마다 호출되는 함수) 에 등록된 OnUpdatePlayer 를 통해 변경된 위치로 이동시킴
